src/hostapi/sndio/pa_sndio.c

```diff
@@ -0,0 +1,767 @@
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+/*
+ * Indentation is done via the .clang-format at the root of the repo
+ * and the following command:
+ * clang-format -style=file -assume-filename=.clang-format -i src/hostapi/sndio/pa_sndio.c
+ */
+#include <errno.h>
+#include <poll.h>
+#include <pthread.h>
+#include <sndio.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+
+#include "pa_allocation.h"
+#include "pa_debugprint.h"
+#include "pa_hostapi.h"
+#include "pa_process.h"
+#include "pa_stream.h"
+#include "pa_util.h"
+
+/*
+ * per-stream data
+ */
+typedef struct PaSndioStream
+{
+    PaUtilStreamRepresentation base;
+    PaUtilBufferProcessor bufproc; /* format conversion */
+    struct sio_hdl *hdl; /* handle for device i/o */
+    struct sio_par par; /* current device parameters */
+    unsigned mode; /* SIO_PLAY, SIO_REC or both */
+    int stopped; /* stop requested or not started */
+    int active; /* thread is running */
+    unsigned long long realpos; /* frame number h/w is processing */
+    char *rbuf, *wbuf; /* bounce buffers for conversions */
+    unsigned long long rpos, wpos; /* bytes read/written */
+    pthread_t thread; /* thread of the callback interface */
+} PaSndioStream;
+
+/*
+ * api "class" data, common to all streams
+ */
+typedef struct PaSndioHostApiRepresentation
+{
+    PaUtilHostApiRepresentation base;
+    PaUtilStreamInterface callback;
+    PaUtilStreamInterface blocking;
+    /*
+     * sndio has no device discovery mechanism and PortAudio has
+     * no way of accepting raw device strings from users.
+     * Normally we just expose the default device, which can be
+     * changed via the AUDIODEVICE environment variable, but we
+     * also allow specifying a list of up to 16 devices via the
+     * PA_SNDIO_AUDIODEVICES environment variable.
+     *
+     * Example:
+     * PA_SNDIO_AUDIODEVICES=default:snd/0.monitor:snd@remote/0
+     */
+#define PA_SNDIO_AUDIODEVICES_MAX 16
+    PaDeviceInfo device_info[PA_SNDIO_AUDIODEVICES_MAX];
+    PaDeviceInfo *infos[PA_SNDIO_AUDIODEVICES_MAX];
+    char *audiodevices;
+} PaSndioHostApiRepresentation;
+
+/*
+ * callback invoked when blocks are processed by the hardware
+ */
+static void sndioOnMove( void *addr, int delta )
+{
+    PaSndioStream *sndio_stream = (PaSndioStream *)addr;
+
+    sndio_stream->realpos += delta;
+}
+
+/*
+ * convert PA encoding to sndio encoding, return true on success
+ */
+static int sndioSetFmt( struct sio_par *par, PaSampleFormat fmt )
+{
+    switch( fmt & ~paNonInterleaved )
+    {
+    case paInt32:
+        par->sig = 1;
+        par->bits = 32;
+        break;
+    case paInt24:
+        par->sig = 1;
+        par->bits = 24;
+        par->bps = 3; /* paInt24 is packed format */
+        break;
+    case paInt16:
+    case paFloat32:
```

Comment:
I'm not sure whether this is a bug, but ideally float32 should be mapped to 24 or 32 bit IO.

Reply:
Also please leave a comment explaining why it works even though it may look odd. That will save you from having people ask questions about the code in the future.

Reply:
Looks like this was a bug I inherited, it doesn't looks like sndio supports floats so I removed the `case paFloat32:` line.

Reply:
Mapped it to 32bit IO, not entirely sure what would be the best way to test or be sure which to pick.